pkg/gadgets/internal/tchandler/tracer.go

```diff
@@ -0,0 +1,282 @@
+	}
+	a.dispatcher.Close()
+}
+
+type Handler struct {
+	// dispatcher map is a program array map with a single element that is used by the
+	// dispatcher to perform a tail call to the gadget program.
+	dispatcherMap *ebpf.Map
+	// key: network interface name on the host side
+	// value: attachment
+	attachments map[string]*attachment
+
+	// socket to talk to netlink
+	tcnl *tc.Tc
+
+	direction AttachmentDirection
+}
+
+func NewHandler(direction AttachmentDirection) (*Handler, error) {
+	var err error
+	var tcnl *tc.Tc
+
+	// We need to create the client on the host network namespace, otherwise it's not able to
+	// create the qdisc and filters.
+	err = netnsenter.NetnsEnter(1, func() error {
+		// Setup tc socket for communication with the kernel
+		tcnl, err = tc.Open(&tc.Config{})
+		if err != nil {
+			return fmt.Errorf("opening rtnetlink socket: %w", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	t := &Handler{
+		attachments: make(map[string]*attachment),
+		tcnl:        tcnl,
+		direction:   direction,
+	}
+	defer func() {
+		if err != nil {
+			t.Close()
+		}
+	}()
+
+	// Keep in sync with tail_call map in bpf/dispatcher.bpf.c
+	dispatcherMapSpec := ebpf.MapSpec{
+		Name:       tailCallMapName,
+		Type:       ebpf.ProgramArray,
+		KeySize:    4,
+		ValueSize:  4,
+		MaxEntries: 1,
+	}
+	t.dispatcherMap, err = ebpf.NewMap(&dispatcherMapSpec)
+	if err != nil {
+		return nil, fmt.Errorf("creating tail call map: %w", err)
+	}
+	return t, nil
+}
+
+func (t *Handler) AttachProg(prog *ebpf.Program) error {
+	return t.dispatcherMap.Update(uint32(0), uint32(prog.FD()), ebpf.UpdateAny)
+}
+
+func (t *Handler) newAttachment(pid uint32, iface *net.Interface, netns uint64, direction AttachmentDirection) (_ *attachment, err error) {
+	a := &attachment{
+		users: map[uint32]struct{}{pid: {}},
+	}
+	defer func() {
+		if err != nil {
+			t.closeAttachment(a)
+		}
+	}()
+
+	dispatcherSpec, err := loadDispatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	consts := map[string]interface{}{
+		"current_netns": uint32(netns),
+	}
+	if err := dispatcherSpec.RewriteConstants(consts); err != nil {
+		return nil, fmt.Errorf("RewriteConstants while attaching to pid %d: %w", pid, err)
+	}
+
+	// We create the clsact qdisc and leak it. We can't remove it because we'll break any other
+	// application (including other ig instances) that are using it.
+	if _, err = createClsActQdisc(t.tcnl, iface); err != nil {
+		return nil, fmt.Errorf("creating clsact qdisc: %w", err)
+	}
+
+	optsIngress := ebpf.CollectionOptions{
+		MapReplacements: map[string]*ebpf.Map{
+			tailCallMapName: t.dispatcherMap,
+		},
+	}
+	if err = dispatcherSpec.LoadAndAssign(&a.dispatcher, &optsIngress); err != nil {
+		return nil, fmt.Errorf("loading ebpf program: %w", err)
+	}
+
+	a.filter, err = addTCFilter(t.tcnl, a.dispatcher.IgNetDisp, iface, direction)
+	if err != nil {
+		return nil, fmt.Errorf("attaching ebpf program to interface %s: %w", iface.Name, err)
+	}
```

Comment:
In these two error cases, you can indeed delete the clsact qdisc. Probably by adding it to the above defer.